mgr/dashboard: add service management for oauth2-proxy

Pegonzal opened the pull request:
Fixes: https://tracker.ceph.com/issues/67651




[screen-capture (24).webm](https://github.com/user-attachments/assets/a6aefe30-1278-4e52-90c1-30eec4c57137)

Video's spec file:
```
[ceph: root@ceph-node-00 /]# ceph orch ls --service_name oauth2-proxy --export
service_type: oauth2-proxy
service_name: oauth2-proxy
placement:
  count: 1
spec:
  allowlist_domains:
  - domain1.com
  - 192.168.1.1:433
  - 192.168.100.100
  - ceph-node-00
  client_id: clientid
  client_secret: clientsecret
  oidc_issuer_url: https://0.0.0.0
  provider_display_name: My OIDC provider
```

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

cloudbehl commented:
Hi Pedro, can you add basic description for Display name, Client ID, Secret , issuer url.
Also Client secret shouldn't  it be type password and hidden?

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -420,7 +438,45 @@ export class ServiceFormComponent extends CdForm implements OnInit {
+          return !(addressTest && portTest);
+        })
+      ]],
+      provider_display_name: [null, [
+        CdValidators.requiredIf({
+          service_type: 'oauth2-proxy'
+        })
```
we can have default value here "My OIDC provider"

Pegonzal replied:
Not a fan of arbitrary default values. 
But I can include it of course

Pegonzal replied:
Added it

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -420,7 +438,45 @@ export class ServiceFormComponent extends CdForm implements OnInit {
-      grafana_admin_password: [null]
+      grafana_admin_password: [null],
+      // oauth2-proxy
+      https_address: [null, [
+        CdValidators.requiredIf({
+          service_type: 'oauth2-proxy'
```
I think this is an optional field and we can move it below in that case

Pegonzal replied:
Sure

afreen23 commented:
> Hi Pedro, can you add basic description for Display name, Client ID, Secret , issuer url. Also Client secret shouldn't it be type password and hidden?

agree we can have client secret hidden , password type

afreen23 commented:
Also, shall we tell user that authetication is required to be enabled in "mgmt-gateway" for this ?

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -50,6 +50,8 @@ export class ServiceFormComponent extends CdForm implements OnInit {
   readonly INGRESS_SUPPORTED_SERVICE_TYPES = ['rgw', 'nfs'];
   readonly SMB_CONFIG_URI_PATTERN = /^(http:|https:|rados:|rados:mon-config-key:)/;
+  readonly OAUTH2_HTTPS_ADDRES_PATTERN = /^((\d{1,3}\.){3}\d{1,3}|([a-zA-Z0-9-_]+\.)*[a-zA-Z0-9-_]+)/;
+  readonly OAUTH2_HTTPS_PORT_PATTERN = /^(6553[0-5]|655[0-2]\d|65[0-4]\d{2}|6[0-4]\d{3}|[1-5]\d{4}|\d{1,4})$/;
```
Beucase the input field takes the IP address and the port, in the same text field.
I did it this way to keep consistency with the CLI (on the Cephadm side it takes an input like `0.0.0.0:4180` all together)

afreen23 replied:
That sounds fine, we can still make this simpler by adding a simple test (see below). You can think about it.

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -617,7 +673,21 @@ export class ServiceFormComponent extends CdForm implements OnInit {
+                'client_id',
+                'client_secret',
+                'oidc_issuer_url'
+              ];
```
Right, I forgot it. Including it

Pegonzal commented:
> Also, shall we tell user that authetication is required to be enabled in "mgmt-gateway" for this ?

I can include that in a `cd-alert` info message maybe

Pegonzal commented:
> Hi Pedro, can you add basic description for Display name, Client ID, Secret , issuer url.
> Also Client secret shouldn't it be type password and hidden?

Done

Pegonzal commented:
jenkins test make check

Pegonzal commented:
jenkins test api

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                     type="password"
+                     formControlName="client_secret"
+                     placeholder="proxy">
+              <cd-help-text i18n>Along with Client ID authenticates application to the IDP</cd-help-text>
+              <span class="invalid-feedback"
```
```suggestion
              <cd-help-text i18n>The client secret for authenticating with the identity provider.</cd-help-text>
              <span class="invalid-feedback"
```

Pegonzal replied:
Using IdP to avoid repetition

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                     class="form-control"
+                     type="text"
+                     formControlName="oidc_issuer_url">
+              <cd-help-text i18n>IDP's issuer url</cd-help-text>
```
```suggestion
              <cd-help-text i18n>The URL of the OpenID Connect (OIDC) issuer</cd-help-text>
```

afreen23 replied:
We should  a basic URL validation here too

Pegonzal replied:
For the URL validation we could check that there is a valid schema present, but a part from that, what other validation could be added?

Pegonzal replied:
Added url validation

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                     type="text"
+                     formControlName="https_address"
+                     placeholder="0.0.0.0:4180">
+              <cd-help-text i18n>Address format [IP|hostname]:port</cd-help-text>
```
```suggestion
              <cd-help-text i18n>The address for HTTPS connections, formatted as 'host:port</cd-help-text>
```

Pegonzal replied:
Left it as [IP|Hostname]:port as it can be either of one them, it will depend on the user's setup

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                    i18n>This field is required.</span>
+              <span class="invalid-feedback"
+                    *ngIf="serviceForm.showError('https_address', frm, 'httpsAddressPattern')"
+                    i18n>Format must be [IP|hostname]:port and port between 0 and 65535</span>
```
```suggestion
                    i18n>Format must be host:port and port between 0 and 65535</span>
```

Pegonzal replied:
I saw this mentioned multiple times but I don't really see the reason for it, this could be either be an IP or a hostname

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -420,7 +438,58 @@ export class ServiceFormComponent extends CdForm implements OnInit {
+            }
+            const [address, port] = value.split(':');
+            const addressTest = this.OAUTH2_HTTPS_ADDRES_PATTERN.test(address);
+            const portTest = this.OAUTH2_HTTPS_PORT_PATTERN.test(port);
```
```suggestion
            const portTest = port >= 0 && port <= 65535;
```

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                     type="text"
+                     formControlName="redirect_url"
+                     placeholder="https://<your-ip>:4180/oauth2/callback">
+              <cd-help-text i18n>The URL oauth2-proxy will redirect to after a successful login.</cd-help-text>
```
```suggestion
              <cd-help-text i18n>The URL <pre>oauth2-proxy</pre> will redirect to after a successful login.</cd-help-text>
```

Pegonzal replied:
Any reason for this?

afreen23 replied:
`oauth2-proxy` is the name of the service, or add "service" after oauth2-proxy perhaps.

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+          <div class="form-group row">
+            <label class="cd-col-form-label required"
+                   for="oidc_issuer_url">
+              <span i18n>Issuer url</span>
```
```suggestion
              <span i18n>OIDC Issuer URL</span>
```

Pegonzal replied:
Addressed it

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+          <div class="form-group row">
+            <label class="cd-col-form-label required"
+                   for="provider_display_name">
+              <span i18n>Provider display name</span>
```
This is how see in service form hence ti be consistent
```suggestion
              <span i18n>Provider Display Name</span>
```

Pegonzal replied:
Following https://carbondesignsystem.com/components/text-input/usage/#content

> Use sentence-style capitalization for all labels, except for product names and proper nouns.

So I left it as it is

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+          <div class="form-group row">
+            <label class="cd-col-form-label required"
+                   for="client_secret">
+              <span i18n>Client secret</span>
```
```suggestion
              <span i18n>Client Secret</span>
```

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+          <div class="form-group row">
+            <label class="cd-col-form-label"
+                   for="https_address">
+              <span i18n>Https address</span>
```
```suggestion
              <span i18n>Https Address</span>
```

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.ts:
```diff
@@ -420,7 +438,58 @@ export class ServiceFormComponent extends CdForm implements OnInit {
+        null,
+        [
+          CdValidators.custom('httpsAddressPattern', (value: string) => {
+            if (_.isEmpty(value)) {
+              return false;
```
Lets create a separate function for this `oauthAddressTest()` .

Pegonzal replied:
Don't really see the appeal for it, tbh

afreen23 replied:
The reason is to maintain code readability and maintenance.

Pegonzal replied:
Added the function

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                     class="form-control"
+                     type="text"
+                     formControlName="redirect_url"
+                     placeholder="https://<your-ip>:4180/oauth2/callback">
```
```suggestion
                     placeholder="https://<IP address>:4180/oauth2/callback">
```

Pegonzal replied:
Left it as <IP|Hostname> as the desired input here is the domain that is either the IP or hostname, depending on user's config

afreen23 replied:
I am confused, as here we are saying `<your-ip>` which means IP address.
This is different from the above :/

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                     class="form-control"
+                     type="password"
+                     formControlName="client_secret"
+                     placeholder="proxy">
```
```suggestion
                     placeholder="">
```
We should have a toggle like show/hide as in password

Pegonzal replied:
Good point, added it

afreen23 reviewed (changes requested):
Thanks @Pegonzal this looks fine, just some cosmetics

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -883,8 +890,121 @@
+                     type="text"
+                     formControlName="client_id"
+                     placeholder="oauth2-proxy">
+              <cd-help-text i18n>Identifies the application to the IDP</cd-help-text>
```
using IdP to avoid repetition

rkachach reviewed (changes requested):
`allowlist_domains: ` is a YAML list. Current code is just putting whatever the field has (comma separated values) in the spec which leads later to crash in cephadm. This field must be a list of strings. In Yaml it looks like:

```
  whitelist_domains:
  - 192.168.100.1:8080
  - domain1.com
```

@Pegonzal plz keep in mind that the field name as changed: https://github.com/ceph/ceph/pull/59766 ... the new name is `allowlist_domains`

Pegonzal commented:
> `allowlist_domains: ` is a YAML list. Current code is just putting whatever the field has (comma separated values) in the spec which leads later to crash in cephadm. This field must be a list of strings. In Yaml it looks like:
> 
> ```
>   whitelist_domains:
>   - 192.168.100.1:8080
>   - domain1.com
> ```
> 
> @Pegonzal plz keep in mind that the field name as changed: #59766 ... the new name is `allowlist_domains`

Updated it @rkachach 
Also updated the video and PR description with spec file

Pegonzal commented:
Jenkins test api

rkachach commented:
Testing the last code I can see that now the spec is generated correctly. Just a small note: spaces must be eliminated from the `whitelist_domains` (in case the user introduces some of them):

```
[ceph: root@ceph-node-0 /]# ceph orch ls --service_name oauth2-proxy --export
service_type: oauth2-proxy
service_name: oauth2-proxy
placement:
  count: 1
  hosts:
  - ceph-node-2
spec:
  allowlist_domains:
  - domain1.com
  - ' 192.168.100.1:8080'
  - 192.168.100.101
  - ceph-node-1
  client_id: oauth2-proxy
  client_secret: vwCLTITb2rtem583523uXcuSNItDsz1I
  oidc_issuer_url: http://192.168.100.1:8080/realms/ceph
  provider_display_name: Redo OIDC provider

```

Pegonzal commented:
> Testing the last code I can see that now the spec is generated correctly. Just a small note: spaces must be eliminated from the `whitelist_domains` (in case the user introduces some of them):
> 
> ```
> [ceph: root@ceph-node-0 /]# ceph orch ls --service_name oauth2-proxy --export
> service_type: oauth2-proxy
> service_name: oauth2-proxy
> placement:
>   count: 1
>   hosts:
>   - ceph-node-2
> spec:
>   allowlist_domains:
>   - domain1.com
>   - ' 192.168.100.1:8080'
>   - 192.168.100.101
>   - ceph-node-1
>   client_id: oauth2-proxy
>   client_secret: vwCLTITb2rtem583523uXcuSNItDsz1I
>   oidc_issuer_url: http://192.168.100.1:8080/realms/ceph
>   provider_display_name: Redo OIDC provider
> ```

Thanks @rkachach for the review. I'll address the issue in the PR (#58618) that will follow this one as I'll need to rebase it after meging this one anyways.
